source/modulo_utils/modulo_utils/binary_io.py

```diff
@@ -0,0 +1,119 @@
+import os
+import struct
+import time
+from datetime import datetime
+from io import BufferedWriter
+from typing import List
+
+import clproto
+from modulo_interfaces.msg import EncodedState
+
+
+class BinaryRecorder():
+    def __init__(self, filepath: str):
+        """
+        Construct the BinaryRecorder. By calling first open() and then write(msg) successively, the BinaryRecorder will
+        write the provided message as binary data to a file under the desired path. 
+
+        :param filepath: The full path of the recording file
+        """
+        self._filepath = filepath
+        if not self._filepath.endswith(".bin"):
+            self._filepath += ".bin"
+        self._file: BufferedWriter
+
+    def open(self):
+        """
+        Open the file for writing.
+        """
+        os.makedirs(os.path.dirname(self._filepath), exist_ok=True)
+        self._file = open(self._filepath, 'wb')
+
+    def write(self, msg: EncodedState):
+        """
+        Write an EncodedState to the file.
+        """
+        self._file.write(clproto.pack_fields([struct.pack('d', time.time()), bytes(msg.data)]))
+
+    def close(self):
+        """
+        Close the file after writing.
+        """
+        self._file.close()
```

Comment:
it would be nice to add an enter and exit method to the class to use with context

```python
    def __enter__(self):
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
```